x/incentive/keeper/refund_tx_decorator.go

```diff
@@ -0,0 +1,64 @@
+package keeper
+
+import (
+	"github.com/babylonlabs-io/babylon/x/incentive/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+var _ sdk.PostDecorator = &RefundTxDecorator{}
+
+type RefundTxDecorator struct {
+	k *Keeper
+}
+
+// NewRefundTxDecorator creates a new RefundTxDecorator
+func NewRefundTxDecorator(k *Keeper) *RefundTxDecorator {
+	return &RefundTxDecorator{
+		k: k,
+	}
+}
+
+func (d *RefundTxDecorator) PostHandle(ctx sdk.Context, tx sdk.Tx, simulate, success bool, next sdk.PostHandler) (sdk.Context, error) {
+	// refund only when finalizing a block or simulating the current tx
+	if ctx.ExecMode() != sdk.ExecModeFinalize && !simulate {
+		return next(ctx, tx, simulate, success)
+	}
+	// avoid refunding failed tx
+	if !success {
+		return next(ctx, tx, simulate, success)
+	}
+	// ignore tx that does not need to pay fee
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return next(ctx, tx, simulate, success)
+	}
+
+	refundableMsgHashList := make([][]byte, 0)
+
+	// iterate over all messages in the tx, and record whether they are refundable
+	for _, msg := range tx.GetMsgs() {
+		msgHash := types.HashMsg(msg)
+		if d.k.HasRefundableMsg(ctx, msgHash) {
```

Comment:
I think we need to filter our the duplicated form the message list or do refunds only in case when there is 1 message in transaction. 

Consider case, if finality provider woud create transactions with finality votes:
1. First vote would be valid finality vote which would be indexable as to rewards
2. All other votes would be duplicates

Transaction would be successful as duplicated vote is not an error:
```
	if err == nil && existingSig.Equals(req.FinalitySig) {
		ms.Logger(ctx).Debug("Received duplicated finiality vote", "block height", req.BlockHeight, "finality provider", req.FpBtcPk)
		// exactly same vote already exists, return success to the provider
		return &types.MsgAddFinalitySigResponse{}, nil
	}
```
Not finality provider can bloat the ledgers for free i.e he can send transactions with 100 duplicated votes and not pay anything.

Reply:
In this example, upon duplicated finality signature the function will return directly without calling `IndexRefundableMsg`, so [this condition](https://github.com/babylonlabs-io/babylon/blob/6e5210d0eacca6570860152d92ec58f4da7c8cf3/x/incentive/keeper/refund_tx_decorator.go#L47-L48) won't be met and the tx won't get refunded. Or did I miss anything?

Reply:
Per offline discussion, this is an issue: the problem is that these messages share the same hash and then all of them are considered refundable in the posthandler. Two ways of addressing this issue:
deduplicate hashes in posthandler before comparing the length
return error upon duplicated finality signature